builtin/game/falling.lua

```diff
@@ -465,23 +479,34 @@ function core.check_single_for_falling(p)
 			if same and d_bottom.paramtype2 == "leveled" and
 					core.get_node_level(p_bottom) <
 					core.get_node_max_level(p_bottom) then
-				convert_to_falling_node(p, n)
-				return true
+				return convert_to_falling_node(p, n)
 			end
 			-- Otherwise only if the bottom node is considered "fall through"
 			if not same and
 					(not d_bottom.walkable or d_bottom.buildable_to) and
 					(core.get_item_group(n.name, "float") == 0 or
 					d_bottom.liquidtype == "none") then
-				convert_to_falling_node(p, n)
-				return true
+				return convert_to_falling_node(p, n)
 			end
 		end
 	end
 
 	if core.get_item_group(n.name, "attached_node") ~= 0 then
+		local def = core.registered_nodes[n.name]
 		if not builtin_shared.check_attached_node(p, n) then
+			if def and def.on_detach then
+				local ok = def.on_detach(table.copy(p), table.copy(n))
+				if (type(ok) == "boolean") then
+					return ok
+				end
+			end
+			local oldmeta = core.get_meta(p):to_table().fields
+
 			drop_attached_node(p)
+
+			if def and def.after_detach then
+				def.after_detach(p, n, oldmeta)
```

Comment:
```suggestion
				def.after_detach(vector.new(p), n, oldmeta)
```

(Also: Giving the oldmeta here seems a little bit overkill. A mod could still get it by saving it in a variable in `on_detach`, (which might be hacky).)